app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -96,42 +100,108 @@ public void preStart() throws Exception {
      * @param wakeUpBy The time we want to be sure we wake up by.
      * @return The time until we should wake up. Always between 0 and (approximately) {@code wakeUpBy}.
      */
-    /* package private */ FiniteDuration timeUntilNextTick(final Instant wakeUpBy) {
+    /* package private */ Optional<FiniteDuration> timeUntilExtraTick(final Instant wakeUpBy) {
         final FiniteDuration delta = Duration.fromNanos(ChronoUnit.NANOS.between(_clock.instant(), wakeUpBy));
         if (delta.lt(Duration.Zero())) {
-            return Duration.Zero();
+            return Optional.of(Duration.Zero());
         } else if (delta.lt(TICK_INTERVAL)) {
-            return delta;
-        } else {
-            return TICK_INTERVAL;
+            return Optional.of(delta);
+        }
+        return Optional.empty();
+    }
+
+    private void scheduleExtraTickIfNecessary(final Instant wakeUpBy) {
+        timeUntilExtraTick(wakeUpBy).ifPresent(timeRemaining -> {
+            timers().startSingleTimer(EXTRA_TICK_TIMER_NAME, Tick.INSTANCE, timeRemaining);
+        });
+    }
+
+    private void killSelf() {
+        // TODO(spencerpearson): Per https://doc.akka.io/docs/akka/2.5.4/java/cluster-sharding.html#remembering-entities ,
+        //   might we need to send a Passivate message to our parent?
+        getSelf().tell(PoisonPill.getInstance(), getSelf());
+    }
+
+    /**
+     * Initializes the actor with the given JobRef (if uninitialized), or ensure the the given ref equals the one already initialized with.
+     *
+     * @param ref The JobRef to initialize with.
+     * @return The {@link CachedJob} the actor is initialized to.
+     *   (Guaranteed to equal {@code _cachedJob.get()}; this return value is purely for the typechecker's sake.)
+     * @throws IllegalStateException If the actor was already initialized with a different JobRef.
+     */
+    private CachedJob<T> initializeOrEnsureRefMatch(final JobRef<T> ref) throws IllegalStateException {
+        if (!_cachedJob.isPresent()) {
+            LOGGER.info()
+                    .setMessage("initializing")
+                    .addData("ref", ref)
+                    .log();
+            _cachedJob = Optional.of(new CachedJob<>(_injector, ref));
+        }
+
+        final JobRef<T> oldRef = _cachedJob.get().getRef();
+        if (!oldRef.equals(ref)) {
+            LOGGER.error()
+                    .setMessage("JobRef in received message does not match cached JobRef")
+                    .addData("cached", oldRef)
+                    .addData("new", ref)
+                    .log();
+            killSelf();
```

Comment:
Instead of logging here and killing can we throw an `IllegalStateException`? We can then log the exception from the `preRestart`. This way I feel the control flow is simpler and less explicit. We also have the opportunity to log an exception instead of just a message which should have other integration paths in the future. The only downside I see is losing some the structured nature in the logs; however, since these should _never_ happen, I think it's a reasonable trade-off.